src/ip.rs

```diff
@@ -410,13 +482,93 @@ impl TcpPacket {
     }
 }
 
+/// Errors that can happen while parsing a UDP packet
+#[derive(Debug, Error)]
+pub enum UdpParseErr {
+    /// Size in header is too large for passed in slice
+    #[error("Size in header did not match size of data")]
+    SizeMismatch,
+    /// The slice is less than 8 bytes, the minimum UDP packet size
+    #[error("Passed in data is not large enough for UDP header")]
+    MissingHeader,
+}
+
+/// UDP header and associated data
+pub struct UdpPacket {
```

Comment:
maybe more correct to call it a datagram? since I think the whole thing is an ip packet and this is just a fragment of it